### PR TITLE
Post-processing agent policies

### PR DESCRIPTION
**Pull request by omniscientoctopus**
The goal is to create tools for analyzing agent behaviour. The post-processing is done in two steps:

### 1. Record rollouts

We record rollouts using a recorder so that it is agonostic to the solver, and can be saved to disk etc. (credits to @moratodpg for the suggestion).

```python
from imp_act import make
from imp_act.environments.recorder import Recorder

env = make("ToyExample-v2")
recorded_env = Recorder(env)

# record rollouts
for ep in range(10):

    obs = recorded_env.reset()
    done = False
    while not done:

        # insert your policy here
        actions = [[0] * len(e["road_edge"].segments) for e in env.graph.es]

        obs, reward, done, info = recorded_env.step(actions)
```

The rollouts are recorded on a `tape` (dict)
```python
tape = recorded_env.tape

# can be converted to df for manipulation/saving
df = recorded_env.tape_to_df()
```

### 2.Plotting

```python
from imp_act.environments.rollout_plotter import RolloutPlotter

plotter = RolloutPlotter(env)

ep = 2 # episode to plot
plotter.plot(episode_data=tape[ep])
```

#### Plot 1: Deterioration Process

image
 
#### Plot 2: Travel time and rewards
image

#### Plot 3: Traffic data
image

Currently, we are only analysing a single rollout. In the next steps, we should analyse multiple rollouts to measure aggregate performance etc.

### CODE REVIEW

**Comment by omniscientoctopus** [comment]
You can try it out here: [sandbox.ipynb](https://github.com/AI-for-Infrastructure-Management/imp-act/blob/history-recorder/sandbox.ipynb). I will delete this notebook before merging

File: imp_act/environments/README.md
Comment (DanielHettegger) [inline imp_act/environments/README.md:17]: Would be cool to include this in the main package init. (Similar to "make")

File: imp_act/environments/rollout_plotter.py
Comment (DanielHettegger) [inline imp_act/environments/rollout_plotter.py:150]: This does not seem to be used according to linter.

**Review by DanielHettegger — CHANGES_REQUESTED** [review]
Hey Prateek, thanks for this, I really like the clean implementation as a wrapper. I was also thinking of doing it this way.
I also really like the visualizations you made, these should make inspecting the policies a lot nicer.
I have left some comments on the code. And I think we should merge the budget constraints first, so we should rebase this on one of those branches.

File: imp_act/environments/recorder.py
Comment (omniscientoctopus) [reply imp_act/environments/recorder.py:59]: I tried something like that in the first commit [here](https://github.com/AI-for-Infrastructure-Management/imp-act/commit/6e7e3c90dfc49a30bb872435eae6461cb6d4c79f), but decided to manually add keys because,
- But doing it in this way made accessing the objects inside the 'obs' variable becomes quite cluttered. 
- Some keys such as 'edge_observations' have max_timesteps+1 entries, but 'edge_observations' are not returned by reset, but rather only in 'info', so to access it here, I have to use 'self.env._get_states()'.
 
Overall, I couldn't figure out how to loop over the keys in a "clean" way.

File: imp_act/environments/recorder.py
Comment (omniscientoctopus) [reply imp_act/environments/recorder.py]: I agree. I went with "rollout_data" since I already used "episode_data" in in rollout_plotter.py for something else

File: imp_act/environments/README.md
Comment (omniscientoctopus) [reply imp_act/environments/README.md]: that's a good idea, done! I also included RolloutPlotter.

File: imp_act/environments/recorder.py
Comment (DanielHettegger) [reply imp_act/environments/recorder.py:59]: grafik

What do you think about something like this?
This way the only naming dependency is on the edge_states and the rest is updated automatically should we change the observations.

I can push the changes if you agree.

Comment (omniscientoctopus) [reply imp_act/environments/recorder.py:59]: yes, I think this is much better!

One last thing, we included the "adjacency_matrix" in obs for the case when graph connectivity changes, right? If that feature is not implemented right now, does it makes sense to remove it from the obs dict? This way we can avoid the "record_exclude_keys". Because in the case when connectivity changes, it must recorded?

Comment (DanielHettegger) [reply imp_act/environments/recorder.py]: Cool, thanks Prateek :) 
I think I would leave it as part of the observations, it might already be usefull to the agent to determine the redundancies. And for the challenge the agents should only rely on what is passed using the observation dict, (if the object can be accessed the agent could change rewards, see states etc. ) so it would be best to keep it in.

**Review by DanielHettegger — COMMENTED** [review]
Looks good! Lets wait for the budget branch to merge first and then apply this one.

**Review by DanielHettegger — DISMISSED** [review]

**Comment by DanielHettegger** [comment]
Actually we wanted to remove the Jupyter notebook first. Let's do that after Pablo has had a look.

**Comment by omniscientoctopus** [comment]
Here are the revised plots after the budget feature merge. These plots are under the do-nothing policy. Let me know if we should add any other details to the plots

#### Plot 1: Deterioration Process

image

#### Plot 2: Travel time and rewards

image

#### Plot 3: Budget Remaining

image

#### Plot 4: Traffic Data

image

**Review by moratodpg — DISMISSED** [review]
Very nice addition! It works well on my end. Let's merge it. :)

Some ideas for the future:
+ It'd be nice to include a method for resetting `self.rollout_data` and `self.episode` if needed. 
+ We can also show how to use the plotter within the readme.

**Comment by omniscientoctopus** [comment]
> Some ideas for the future:
> 
> It'd be nice to include a method for resetting self.rollout_data and self.episode if needed.
We can also show how to use the plotter within the readme.

@moratodpg 
These were easy to implement, so I have added them already.

**Review by moratodpg — DISMISSED** [review]
Well done, @omniscientoctopus.  I can see now how important "edge 9" is. :)

All working well from my end.

**Review by moratodpg — APPROVED** [review]
💯

**Pull request merged**
